### PR TITLE
[FABG-736] Supported java chaincode

### PR DESCRIPTION
**Pull request by yakumioto**
Supported install, instantiate and upgrade of java chaincode

Signed-off-by: yakumioto <yaku.mioto@gmail.com>

### CODE REVIEW

File: pkg/client/resmgmt/lscc.go [hunk 1]
```diff
@@ -39,6 +41,7 @@ type chaincodeDeployRequest struct {
 	Name       string
 	Path       string
 	Version    string
+	Lang       string
```
Comment (alikic) [inline pkg/client/resmgmt/lscc.go]: type should be pb.ChaincodeSpec_Type

File: pkg/client/resmgmt/resmgmt.go [hunk 2]
```diff
@@ -872,8 +874,8 @@ func checkRequiredCCProposalParams(channelID string, req InstantiateCCRequest) e
 		return errors.New("must provide channel ID")
 	}
 
-	if req.Name == "" || req.Version == "" || req.Path == "" || req.Policy == nil {
-		return errors.New("Chaincode name, version, path and policy are required")
+	if req.Name == "" || req.Lang == "" || req.Version == "" || req.Path == "" || req.Policy == nil {
+		return errors.New("Chaincode name, language, version, path and policy are required")
```
Comment (alikic) [inline pkg/client/resmgmt/resmgmt.go]: Generally, we try not to break the existing client code. Language should be optional and default to golang.

File: pkg/client/resmgmt/lscc.go (hunk 1 above)
Comment (yakumioto) [reply pkg/client/resmgmt/lscc.go]:
```golang
ccds := &pb.ChaincodeDeploymentSpec{ChaincodeSpec: &pb.ChaincodeSpec{
		Type: pb.ChaincodeSpec_GOLANG, ChaincodeId: &pb.ChaincodeID{Name: chaincode.Name, Path: chaincode.Path, Version: chaincode.Version},
		Type: pb.ChaincodeSpec_Type(pb.ChaincodeSpec_Type_value[strings.ToUpper(chaincode.Lang)]), ChaincodeId: &pb.ChaincodeID{Name: chaincode.Name, Path: chaincode.Path, Version: chaincode.Version},
		Input: &pb.ChaincodeInput{Args: chaincode.Args}}}
```
Hi, there is type conversion here.

File: pkg/client/resmgmt/resmgmt.go (hunk 2 above)
Comment (yakumioto) [reply pkg/client/resmgmt/resmgmt.go]: Hi, has been modified to golang as the default chaincode type.

File: pkg/client/resmgmt/lscc.go (hunk 1 above)
Comment (alikic) [reply pkg/client/resmgmt/lscc.go]: Generally, API should be type safe. Client has to import the protobuf package already, for "CollConfig []*pb.CollectionConfig", so I see no reason for not having type safe interface.

File: pkg/client/resmgmt/resmgmt.go (hunk 2 above)
Comment (alikic) [reply pkg/client/resmgmt/resmgmt.go]: Should also revert the client code (tests) where "golang" was added, we should try not to change old client code if API is not broken. There should be at least one test with "golang" though.

File: pkg/client/resmgmt/lscc.go (hunk 1 above)
Comment (yakumioto) [reply pkg/client/resmgmt/lscc.go]: Thanks for answering.

File: pkg/client/resmgmt/resmgmt.go (hunk 2 above)
Comment (yakumioto) [reply pkg/client/resmgmt/resmgmt.go]: Hi, Unit tests have been added.

File: pkg/client/resmgmt/resmgmt_test.go [hunk 3]
```diff
@@ -1612,6 +1612,42 @@ func TestGetConfigSignaturesFromIdentities(t *testing.T) {
...
 
+func TestCheckRequiredCCProposalParams(t *testing.T) {
+	// Valid request
+	ccPolicy := cauthdsl.SignedByMspMember("Org1MSP")
+	req := InstantiateCCRequest{Name: "name", Version: "version", Path: "path", Policy: ccPolicy}
+
+	// Test empty channel lang
+	err := checkRequiredCCProposalParams("mychannel", &req)
+	if err != nil {
+		t.Fatal("Should have failed for empty request")
```
Comment (alikic) [inline pkg/client/resmgmt/resmgmt_test.go]: Wrong message?

File: pkg/client/resmgmt/example_test.go [hunk 4]
```diff
@@ -64,7 +65,7 @@ func Example() {
 
 	// Instantiate example chaincode on channel 'mychannel'
 	ccPolicy := cauthdsl.SignedByMspMember("Org1MSP")
-	instantiateReq := InstantiateCCRequest{Name: "ExampleCC", Version: "v0", Path: "path", Policy: ccPolicy}
+	instantiateReq := InstantiateCCRequest{Name: "ExampleCC", Version: "v0", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/example_test.go]: Should revert this, as this PR doesn't break the old client code.

File: pkg/client/resmgmt/example_test.go [hunk 5]
```diff
@@ -173,7 +174,7 @@ func ExampleWithTargetFilter() {
 	}
 
 	ccPolicy := cauthdsl.SignedByMspMember("Org1MSP")
-	req := InstantiateCCRequest{Name: "ExampleCC", Version: "v0", Path: "path", Policy: ccPolicy}
+	req := InstantiateCCRequest{Name: "ExampleCC", Version: "v0", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/example_test.go]: Should revert this, as this PR doesn't break the old client code.

File: pkg/client/resmgmt/example_test.go [hunk 6]
```diff
@@ -292,7 +293,7 @@ func ExampleClient_InstantiateCC() {
...
```
Comment (alikic) [inline pkg/client/resmgmt/example_test.go]: Should revert this, as this PR doesn't break the old client code.

File: pkg/client/resmgmt/example_test.go [hunk 7]
```diff
@@ -316,7 +317,7 @@ func ExampleClient_UpgradeCC() {
 	}
 
 	ccPolicy := cauthdsl.SignedByMspMember("Org1MSP")
-	req := UpgradeCCRequest{Name: "ExampleCC", Version: "v1", Path: "path", Policy: ccPolicy}
+	req := UpgradeCCRequest{Name: "ExampleCC", Version: "v1", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/example_test.go]: Should revert this, as this PR doesn't break the old client code.

File: pkg/client/resmgmt/resmgmt.go (hunk 2 above)
Comment (alikic) [reply pkg/client/resmgmt/resmgmt.go]: Great, just need to revert the client code (tests where "golang" option was added) to the original, as this PR is not breaking it. I added comments below.

File: pkg/client/resmgmt/example_test.go (hunk 7 above)
Comment (yakumioto) [reply pkg/client/resmgmt/example_test.go]: Thanks for pointing out, that example_test.go has been reverted.

File: pkg/client/resmgmt/resmgmt_test.go (hunk 3 above)
Comment (yakumioto) [reply pkg/client/resmgmt/resmgmt_test.go]: Thanks, The wrong message has been deleted.

File: pkg/client/resmgmt/resmgmt_test.go [hunk 8]
```diff
@@ -816,7 +816,7 @@ func TestInstantiateCCWithDifferentMSP(t *testing.T) {
 
 	// Valid request
 	ccPolicy := cauthdsl.SignedByMspMember("otherMSP")
-	req := InstantiateCCRequest{Name: "name", Version: "version", Path: "path", Policy: ccPolicy}
+	req := InstantiateCCRequest{Name: "name", Version: "version", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/resmgmt_test.go]: Should revert this, as this PR shouldn't break the old client code.

File: pkg/client/resmgmt/resmgmt_test.go [hunk 9]
```diff
@@ -909,7 +909,7 @@ func TestUpgradeCCWithDifferentMSP(t *testing.T) {
 
 	// Valid request
 	ccPolicy := cauthdsl.SignedByMspMember("otherMSP")
-	req := UpgradeCCRequest{Name: "name", Version: "version", Path: "path", Policy: ccPolicy}
+	req := UpgradeCCRequest{Name: "name", Version: "version", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/resmgmt_test.go]: Should revert this, as this PR shouldn't break the old client code.

File: pkg/client/resmgmt/resmgmt_test.go [hunk 10]
```diff
@@ -967,7 +967,7 @@ func TestCCProposal(t *testing.T) {
 	rc := setupResMgmtClient(t, ctx)
 
 	ccPolicy := cauthdsl.SignedByMspMember("Org1MSP")
-	instantiateReq := InstantiateCCRequest{Name: "name", Version: "version", Path: "path", Policy: ccPolicy}
+	instantiateReq := InstantiateCCRequest{Name: "name", Version: "version", Lang: pb.ChaincodeSpec_GOLANG, Path: "path", Policy: ccPolicy}
```
Comment (alikic) [inline pkg/client/resmgmt/resmgmt_test.go]: Should revert this, as this PR shouldn't break the old client code.

Comment (yakumioto) [reply pkg/client/resmgmt/resmgmt_test.go]: OK, Fixed.

**Pull request merged**